app/grandchallenge/algorithms/models.py

```diff
@@ -69,6 +69,38 @@
 JINJA_ENGINE = sandbox.ImmutableSandboxedEnvironment()
 
 
+class AlgorithmInterfaceManager(models.Manager):
+    def get_existing_interface_for_inputs_and_outputs(
+        self, *, inputs, outputs
+    ):
+        try:
+            return self.annotate(
+                input_count=Count("inputs", distinct=True),
+                output_count=Count("outputs", distinct=True),
+            ).get(
+                inputs__in=inputs,
+                outputs__in=outputs,
+                input_count=len(inputs),
+                output_count=len(outputs),
+            )
+        except ObjectDoesNotExist:
+            return None
+
+
+class AlgorithmInterface(models.Model):
+    inputs = models.ManyToManyField(
+        to=ComponentInterface, related_name="inputs"
+    )
+    outputs = models.ManyToManyField(
+        to=ComponentInterface, related_name="outputs"
+    )
+
+    objects = AlgorithmInterfaceManager()
+
+    def delete(self, *args, **kwargs):
+        raise ValidationError("AlgorithmInterfaces cannot be deleted.")
```

Comment:
What about if a bulk delete is used? Can we catch if the inputs or outputs are changed?

Reply:
To prevent bulk deletion, I can override the `delete` method on the model manager. Catching if inputs and outputs have changed is not possible, I'm afraid. I had mentioned this on the pitch before. The only thing we can do is make them ineditable in the admin (this is what `def has_change_permission(self, request, obj=None): return False` does) and by not allowing these fields to be updated by the user later on through the form (we will have a GetOrCreate view and then updating will be restricted to changing the `is_default` on the through table entry, I'd say. Is that sufficient?

Reply:
Yes probably